lib/connections/service.go

```diff
@@ -336,176 +338,203 @@ func (s *service) handle(ctx context.Context) error {
-						if t, err := s.discoverer.Lookup(ctx, deviceID); err == nil {
-							addrs = append(addrs, t...)
-						}
-					}
-				} else {
-					addrs = append(addrs, addr)
-				}
-			}
+		select {
+		case <-time.After(sleep):
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
 
-			addrs = util.UniqueTrimmedStrings(addrs)
+func (s *service) bestDialerPrio(cfg config.Configuration) int {
+	first := true
+	var bestDialerPrio int
+	for _, df := range dialers {
+		if df.Valid(cfg) != nil {
+			continue
+		}
+		if prio := df.Priority(); first || prio < bestDialerPrio {
+			first = false
+			bestDialerPrio = prio
+		}
+	}
+	return bestDialerPrio
+}
 
-			l.Debugln("Reconnect loop for", deviceID, addrs)
+func (s *service) dialDevices(ctx context.Context, now time.Time, cfg config.Configuration, bestDialerPrio int, nextDial map[string]time.Time, initial bool) {
+	for _, deviceCfg := range cfg.Devices {
+		// Don't attempt to connect to ourselves...
+		if deviceCfg.DeviceID == s.myID {
+			continue
+		}
 
-			dialTargets := make([]dialTarget, 0)
+		// Don't attempt to connect to paused devices...
+		if deviceCfg.Paused {
+			continue
+		}
 
-			for _, addr := range addrs {
-				// Use a special key that is more than just the address, as you might have two devices connected to the same relay
-				nextDialKey := deviceID.String() + "/" + addr
-				seen = append(seen, nextDialKey)
-				nextDialAt, ok := nextDial[nextDialKey]
-				if ok && initialRampup >= sleep && nextDialAt.After(now) {
-					l.Debugf("Not dialing %s via %v as sleep is %v, next dial is at %s and current time is %s", deviceID, addr, sleep, nextDialAt, now)
-					continue
-				}
-				// If we fail at any step before actually getting the dialer
-				// retry in a minute
-				nextDial[nextDialKey] = now.Add(time.Minute)
+		// Don't connect to devices that are already connected using the
+		// best possible dialer priority...
+		connection, connected := s.model.Connection(deviceCfg.DeviceID)
+		if connected && connection.Priority() == bestDialerPrio {
+			continue
+		}
 
-				uri, err := url.Parse(addr)
-				if err != nil {
-					s.setConnectionStatus(addr, err)
-					l.Infof("Parsing dialer address %s: %v", addr, err)
-					continue
-				}
+		dialTargets := s.resolveDialTargets(ctx, now, cfg, deviceCfg, nextDial, initial)
+		if conn, ok := s.dialParallel(ctx, deviceCfg.DeviceID, dialTargets); ok {
+			s.conns <- conn
+		}
+	}
+}
 
-				if len(deviceCfg.AllowedNetworks) > 0 {
-					if !IsAllowedNetwork(uri.Host, deviceCfg.AllowedNetworks) {
-						s.setConnectionStatus(addr, errors.New("network disallowed"))
-						l.Debugln("Network for", uri, "is disallowed")
-						continue
-					}
-				}
+func (s *service) resolveDialTargets(ctx context.Context, now time.Time, cfg config.Configuration, deviceCfg config.DeviceConfiguration, nextDial map[string]time.Time, initial bool) []dialTarget {
+	deviceID := deviceCfg.DeviceID
+	connection, connected := s.model.Connection(deviceID)
```

Comment:
Conn is already obtained just before calling this function and could thus be passed in.